# Wrangle Data
##  Project : We Rate Dogs 

My goal in this wrangling act is to gather, assess and cleaning dataset from 3 different sources to finally be able to consolidate all this data into a single dataset to create interesting and trustworthy analyses and visualizations.

### Table of Content
<ul>
    <li><a href="#Data_Gathering">I. Data Gathering</a></li>
    <ul>
        <li><a href="#twit_arch">1. twitter_archive_enhanced.csv</a></li>
    <li><a href="#img_pred">2.  image_predictions.tsv Programmatically download</a></li>
    <li><a href="#tweet_metrics">3.tweet_json.txt</a></li>
</ul>
    <li><a href="#Data_Assessment">II. Data Assessment</a></li>
    <li><a href="#Data_Cleaning">III.Data Cleaning</a></li>
       <ul>
            <li><a href="#cleaning">1. Cleaning</a></li>
           <ul>
                    <li><a href="#clean1">A. Wrong Datatypes for timestamp column</a></li>
                    <li><a href="#clean2">B. Only original tweet (no retweet or reply)</a></li>
               <li><a href="#clean3">C. Wrong ratings score</a></li>
               <li><a href="#clean4">D. Delete denominator > 10</a></li>
               <li><a href="#clean5">E. Absurd numerator ratings left</a></li>
               <li><a href="#clean6">F. HTML in Source column</a></li>
               <li><a href="#clean7">G. Name column inconsistencies</a></li>
                    <li><a href="#clean8">H. standardization of dog breed names</a></li>
                </ul>      
            <li><a href="#tidy">2. Tydinness</a></li>
                <ul>
                    <li><a href="#tidy1">A. Stage dog column</a></li>
                    <li><a href="#tidy2">B. Delete the retweet and reply columns from archive</a></li>
                    <li><a href="#tidy3">C. Tydinnes issues of image_prediction_clean</a></li>
                </ul>
        </ul>
    <li><a href="#merge">IV. Merging column</a></li>
    <li><a href="#store">V. Store twitter_archive_master</a></li>
</ul>

In [1]:
import pandas as pd
import numpy as np
import requests
import os
import tweepy
import json
import csv

<a id='Data_Gathering'></a>
#                                     I. Data Gathering

<a id='twit_arch'></a>
## 1. twitter_archive_enhanced.csv

twitter_archive_enhanced.csv has been manually uploaded to the notebook

#### Read and visualize twitter-archive-enhanced.csv

In [2]:
twit_arch = pd.read_csv('twitter-archive-enhanced.csv')
twit_arch.head(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None  
2                  10   Archie  None    None   None  None

In [3]:
twit_arch.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

<a id='img_pred'></a>
## 2. image_predictions.tsv Programmatically download

#### Download, Read, visualize image-predictions.tsv

In [4]:
urls = ['https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv']

for url in urls:
    response = requests.get(url)
    with open (os.path.join(url.split('/')[-1]), mode='wb') as file:
        file.write(response.content)
        
img_pred = pd.read_csv('image-predictions.tsv', sep='\t')
img_pred.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

<a id='tweet_metrics'></a>
## 3. tweet_json.txt 

#### Calling & accessing Twitter API using Tweepy

In [5]:
consumer_key = 'zlISSzjLNrpDoNxXz2MiEdCnX'
consumer_secret = 'cjsHFgxdRyRf1lK8mbgkxxFBxGzuNThciGME6RlvAfbKdlkAwj'
access_token = '2729738915-AVLRkroHmQBJuL7egsOXpHyldhhZ9VDw860h54y'
access_secret = 'qnWEA7bV5RMLoSNiY22A0JK0vCQmNNPwGAMOVssianbab'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, 
                 parser = tweepy.parsers.JSONParser(), 
                 wait_on_rate_limit = True, 
                 wait_on_rate_limit_notify = True)

#### Retrieve tweet ID Status, retweet count, and favorite count into tweet_json.txt

In [9]:
tweet_ok = []
tweet_error = []
for tweet_id in twit_arch['tweet_id']:
    try:
        tweet_ok.append(api.get_status(tweet_id, tweet_mode= 'extended'))
    except Exception as e:
         tweet_error.append(tweet_id)

Rate limit reached. Sleeping for: 410
Rate limit reached. Sleeping for: 724
Rate limit reached. Sleeping for: 702


In [13]:
print("The list of tweets" ,len(tweet_ok))
print("The list of tweets no found" , len(tweet_error))

The list of tweets 2331
The list of tweets no found 25


In [23]:
tweet_dicts = []
for each_json_tweet in tweet_ok:
    tweet_dicts.append(each_json_tweet)

In [24]:
with open('tweet_json.txt', 'w') as file:
        file.write(json.dumps(tweet_ok, indent=4))

In [5]:
#identify information of interest from JSON dictionaries in txt file
#and put it in a dataframe called tweet JSON
tweet_stat = []
with open('tweet_json.txt', encoding='utf-8') as json_file:  
    all_tweet = json.load(json_file)
    for each_dictionary in all_tweet:
        tweet_id = each_dictionary['id']
        fav = each_dictionary['favorite_count']
        rt = each_dictionary['retweet_count']

        tweet_stat.append({'tweet_id': str(tweet_id),
                             'fav': int(fav),
                             'rt': int(rt)
                            })
        tweet_metrics = pd.DataFrame(tweet_stat, columns = ['tweet_id', 'fav','rt'])

In [6]:
tweet_metrics.head()

tweet_id    fav    rt
0  892420643555336193  36349  7733
1  892177421306343426  31320  5727
2  891815181378084864  23592  3788
3  891689557279858688  39654  7901
4  891327558926688256  37849  8520

<a id='Data_Assessment'></a>
#                                     II. Data Assessment

In [7]:
twit_arch.head(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None  
2                  10   Archie  None    None   None  None

### Visually :

The following columns have missing value :

    - in_reply_to_status_id
    - in_reply_to_user_id
    - retweeted_status_id
    - retweeted_status_user_id
    - retweeted_status_timestamp

In [8]:
twit_arch.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [9]:
twit_arch[twit_arch.tweet_id.duplicated()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [10]:
twit_arch[twit_arch.tweet_id.duplicated()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [11]:
twit_arch[twit_arch.expanded_urls.duplicated()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
55    881633300179243008           8.816070e+17         4.738443e+07   
64    879674319642796034           8.795538e+17         3.105441e+09   
75    878281511006478336                    NaN                  NaN   
76    878057613040115712                    NaN                  NaN   
98    873213775632977920                    NaN                  NaN   
113   870726314365509632           8.707262e+17         1.648776e+07   
126   868552278524837888                    NaN                  NaN   
135   866450705531457537                    NaN                  NaN   
136   866334964761202691                    NaN                  NaN   
148   863427515083354112           8.634256e+17         7.759620e+07   
150   863062471531167744                    NaN                  NaN   
161   860563773140209665                    NaN                  NaN   
169   859196978902773760                    NaN                  NaN   
179   857214891891077121           8.571567e+17         1.806710e+08   
185   856330835276025856                    NaN                  NaN   
186   856288084350160898           8.562860e+17         2.792810e+08   
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
218   850333567704068097           8.503288e+17         2.195506e+07   
228   848213670039564288           8.482121e+17         4.196984e+09   
234   847617282490613760           8.476062e+17         4.196984e+09   
253   844704788403113984                    NaN                  NaN   
264   842163532590374912                    NaN                  NaN   
271   841077006473256960                    NaN                  NaN   
274   840698636975636481           8.406983e+17         8.405479e+17   
276   840632337062862849                    NaN                  NaN   
280   839549326359670784                    NaN                  NaN   
290   838150277551247360           8.381455e+17         2.195506e+07   
291   838085839343206401           8.380855e+17         2.894131e+09   
293   837820167694528512                    NaN                  NaN   
...                  ...                    ...                  ...   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1605  685681090388975616           6.855479e+17         4.196984e+09   
1618  684969860808454144           6.849598e+17         4.196984e+09   
1655  683391852557561860                    NaN                  NaN   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1683  681694085539872773                    NaN                  NaN   
1688  681523177663676416                    NaN                  NaN   
1689  681340665377193984           6.813394e+17         4.196984e+09   
1726  680055455951884288                    NaN                  NaN   
1742  679462823135686656                    NaN                  NaN   
1774  678023323247357953           6.780211e+17         4.196984e+09   
1806  676936541936185344                    NaN                  NaN   
1819  676590572941893632           6.765883e+17         4.196984e+09   
1844  675849018447167488           6.758457e+17         4.196984e+09   
1895  674742531037511680           6.747400e+17         4.196984e+09   
1905  674606911342424069           6.744689e+17         4.196984e+09   
1914  674330906434379776           6.658147e+17         1.637468e+07   
1940  673716320723169284           6.737159e+17         4.196984e+09   
1970  673295268553605120                    NaN                  NaN   
2038  671550332464455680           6.715449e+17         4.196984e+09   
2149  669684865554620416           6.693544e+17         4.196984e+09   
2180  669000397445533696                    NaN                  NaN   
2189  668967877119254528           6.689207e+17         2.143566e+07   
2212  668587383441514497                    NaN         

In [12]:
twit_arch[twit_arch.expanded_urls.isnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30    886267009285017600           8.862664e+17         2.281182e+09   
55    881633300179243008           8.816070e+17         4.738443e+07   
64    879674319642796034           8.795538e+17         3.105441e+09   
113   870726314365509632           8.707262e+17         1.648776e+07   
148   863427515083354112           8.634256e+17         7.759620e+07   
179   857214891891077121           8.571567e+17         1.806710e+08   
185   856330835276025856                    NaN                  NaN   
186   856288084350160898           8.562860e+17         2.792810e+08   
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
218   850333567704068097           8.503288e+17         2.195506e+07   
228   848213670039564288           8.482121e+17         4.196984e+09   
234   847617282490613760           8.476062e+17         4.196984e+09   
274   840698636975636481           8.406983e+17         8.405479e+17   
290   838150277551247360           8.381455e+17         2.195506e+07   
291   838085839343206401           8.380855e+17         2.894131e+09   
313   835246439529840640           8.352460e+17         2.625958e+07   
342   832088576586297345           8.320875e+17         3.058208e+07   
346   831926988323639298           8.319030e+17         2.068372e+07   
375   828361771580813312                    NaN                  NaN   
387   826598799820865537           8.265984e+17         4.196984e+09   
409   823333489516937216           8.233264e+17         1.582854e+09   
427   821153421864615936           8.211526e+17         1.132119e+08   
498   813130366689148928           8.131273e+17         4.196984e+09   
513   811647686436880384           8.116272e+17         4.196984e+09   
570   801854953262350336           8.018543e+17         1.185634e+07   
576   800859414831898624           8.008580e+17         2.918590e+08   
611   797165961484890113           7.971238e+17         2.916630e+07   
701   786051337297522688           7.727430e+17         7.305050e+17   
707   785515384317313025                    NaN                  NaN   
843   766714921925144576           7.667118e+17         4.196984e+09   
857   763956972077010945           7.638652e+17         1.584641e+07   
967   750381685133418496           7.501805e+17         4.717297e+09   
1005  747651430853525504           7.476487e+17         4.196984e+09   
1080  738891149612572673           7.384119e+17         3.589728e+08   
1295  707983188426153984           7.079801e+17         2.319108e+09   
1345  704491224099647488           7.044857e+17         2.878549e+07   
1445  696518437233913856                    NaN                  NaN   
1446  696490539101908992           6.964887e+17         4.196984e+09   
1474  693644216740769793           6.936422e+17         4.196984e+09   
1479  693582294167244802           6.935722e+17         1.198989e+09   
1497  692423280028966913           6.924173e+17         4.196984e+09   
1523  690607260360429569           6.903413e+17         4.670367e+08   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1605  685681090388975616           6.855479e+17         4.196984e+09   
1618  684969860808454144           6.849598e+17         4.196984e+09   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1689  681340665377193984           6.813394e+17         4.196984e+09   
1774  678023323247357953           6.780211e+17         4.196984e+09   
1819  676590572941893632           6.765883e+17         4.196984e+09   
1844  675849018447167488           6.758457e+17         4.196984e+09   
1895  674742531037511680           6.747400e+17         4.196984e+09   
1905  674606911342424069           6.744689e+17         4.196984e+09   
1914  674330906434379776           6.658147e+17         1.637468e+07   
1940  673716320723169284           6.737159e+17         

In [13]:
twit_arch.name.value_counts()

None         745
a             55
Charlie       12
Cooper        11
Lucy          11
Oliver        11
Tucker        10
Lola          10
Penny         10
Bo             9
Winston        9
Sadie          8
the            8
Toby           7
Daisy          7
Bailey         7
an             7
Buddy          7
Leo            6
Scout          6
Koda           6
Jack           6
Bella          6
Milo           6
Stanley        6
Jax            6
Dave           6
Rusty          6
Oscar          6
Bentley        5
            ... 
Akumi          1
Edmund         1
Pepper         1
Brooks         1
Newt           1
Mojo           1
Charl          1
Kathmandu      1
Bauer          1
Koko           1
BeBe           1
Sunshine       1
Lipton         1
Pip            1
Obi            1
Taco           1
Lolo           1
Ralphé         1
Mike           1
Ralphie        1
Venti          1
Goose          1
Jeb            1
Humphrey       1
Maks           1
Keet           1
Sojourner      1
Karll         

Some Name are inconsistent, they only have small caps

In [14]:
num_aberation = twit_arch.rating_numerator.value_counts()
num_aberation.sort_index(ascending=False)

1776      1
960       1
666       1
420       2
204       1
182       1
165       1
144       1
143       1
121       1
99        1
88        1
84        1
80        1
75        2
60        1
50        1
45        1
44        1
27        1
26        1
24        1
20        1
17        1
15        2
14       54
13      351
12      558
11      464
10      461
9       158
8       102
7        55
6        32
5        37
4        17
3        19
2         9
1         9
0         2
Name: rating_numerator, dtype: int64

Some numerator are way above 10, according to the prject motivation it's seems normal but we'll check if their not corrupt

In [15]:
den_aberation = twit_arch.rating_denominator.value_counts()
den_aberation.sort_index(ascending=False)

170       1
150       1
130       1
120       1
110       1
90        1
80        2
70        1
50        3
40        1
20        2
16        1
15        1
11        3
10     2333
7         1
2         1
0         1
Name: rating_denominator, dtype: int64

Some denominator does not equal 10, we'll have to check those inconsistency

### Programmatically :

- It confirms our prior visual assessment, the `expanded_urls` columns also have missing values.

- The timestamp have the wrong data types (object instead of timestamp)

- While checking the null urls the last 4 columns seems untidy

- There is also a lot of None or non sens value (ex: a, an, bo, the, officially  ...) in name columns.

- Source columns have html in it

- The ratings in numerator and denominator columns are biased (should be 10) 

### `twit_arch` table

#### Missing Value for the Following Columns : 
    - in_reply_to_status_id
    - in_reply_to_user_id
    - retweeted_status_id
    - retweeted_status_user_id
    - retweeted_status_timestamp
    - expanded_urls
    
#### Wrong Data Types :
    - Timestamp
    
#### Cleanliness issue :
    - Check & fix error of rating_numerator
    - The rating_denominator should always be 10
    - HTML in Source column
    - None & non sens value in name columns

##### `img_pred` table

In [16]:
img_pred.head(3)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True

Visually:

We see that the img_pred table is not tidy (p1 : p3 dog)

The p1,p2,p3 column have underscore and mix of capital & small caps

In [17]:
img_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [18]:
img_pred.p1.value_counts()

golden_retriever             150
Labrador_retriever           100
Pembroke                      89
Chihuahua                     83
pug                           57
chow                          44
Samoyed                       43
toy_poodle                    39
Pomeranian                    38
cocker_spaniel                30
malamute                      30
French_bulldog                26
Chesapeake_Bay_retriever      23
miniature_pinscher            23
seat_belt                     22
German_shepherd               20
Staffordshire_bullterrier     20
Siberian_husky                20
web_site                      19
Cardigan                      19
Maltese_dog                   18
Shetland_sheepdog             18
beagle                        18
Eskimo_dog                    18
teddy                         18
Shih-Tzu                      17
Rottweiler                    17
Lakeland_terrier              17
kuvasz                        16
Italian_greyhound             16
          

There is less record than the twit_arch df

Some Breed does not exist in real life such as scorpion, mailbox, pillow ...

Mix between capital and small capps letter


##### `tweet_metrics` table

In [19]:
tweet_metrics.head(3)

tweet_id    fav    rt
0  892420643555336193  36349  7733
1  892177421306343426  31320  5727
2  891815181378084864  23592  3788

In [20]:
tweet_metrics.info(3)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 3 columns):
tweet_id    2331 non-null object
fav         2331 non-null int64
rt          2331 non-null int64
dtypes: int64(2), object(1)
memory usage: 54.7+ KB


#### Quality
### `twit_arch` table

#### Missing Value for the Following Columns : 
    - in_reply_to_status_id
    - in_reply_to_user_id
    - retweeted_status_id
    - retweeted_status_user_id
    - retweeted_status_timestamp
    
    - expanded_urls
    
#### Wrong Data Types :
    - Timestamp
    
#### Cleanliness issue :
    - The max rating_numerator should be 10
    - The rating_denominator should always be 10
    - HTML in Source column
    - None & non sens value in name columns

### `img_pred` table
    - less record than the twit_arch df
    - Some Breed does not exist in real life such as scorpion, mailbox, pillow ...
    - Mix between capital and small capps letter and underscore

#### Tidiness
- The retweet and reply are not interesting since they'll be empty
- doggo, floofer, pupper, puppo should be one column
- The prediction could be only 3 columns instead of 9
- Merge the twit_arch & tweet_metric

<a id='Data_Cleaning'></a>
#                                     III. Data Cleaning & Tydinness

<a id='cleaning'></a>
### I. Cleaning

### `twit_arch` table

In [21]:
twitter_archive_clean = twit_arch.copy()

In [22]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

<a id='clean1'></a>
#### I. Wrong Datatypes for timestamp column

The timestamp column is an object instead of a proper timestamp

In [23]:
twitter_archive_clean.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None

In [24]:
twitter_archive_clean['timestamp'] = pd.to_datetime(twitter_archive_clean['timestamp'])

In [25]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null datetime64[ns]
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: datetime64[ns](1

In [26]:
twitter_archive_clean.head(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   

            timestamp                                             source  \
0 2017-08-01 16:23:56  <a href="http://twitter.com/download/iphone" r...   
1 2017-08-01 00:17:27  <a href="http://twitter.com/download/iphone" r...   
2 2017-07-31 00:18:03  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None  
2                  10   Archie  None    None   None  None

<a id='clean2'></a>
#### II. Only original tweet (no retweet or reply)

As state during the project motivation section we have to delete all the rows how is not null in the following columns :

    - retweeted_status_id
    - retweeted_status_user_id
    - retweeted_status_timestamp

In [27]:
twitter_archive_clean = twitter_archive_clean.query("retweeted_status_id == 'NaN' and in_reply_to_user_id =='NaN'")

In [28]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2097 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     2097 non-null datetime64[ns]
source                        2097 non-null object
text                          2097 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2094 non-null object
rating_numerator              2097 non-null int64
rating_denominator            2097 non-null int64
name                          2097 non-null object
doggo                         2097 non-null object
floofer                       2097 non-null object
pupper                        2097 non-null object
puppo                         2097 non-null object
dtypes: datetime64[ns](1), float

<a id='clean3'></a>
#### III. Wrong ratings score

Even if the project motivation specifies that's The fact that the rating numerators are greater than the denominators does not need to be cleaned some rating seems really overrated and need a check

In [29]:
twitter_score_status = twitter_archive_clean[['tweet_id','text','rating_numerator','rating_denominator']]
pd.set_option('display.max_colwidth', -1)
twitter_score_status.query("rating_denominator != 10")

tweet_id  \
433   820690176645140481   
516   810984652412424192   
902   758467244762497024   
1068  740373189193256964   
1120  731156023742988288   
1165  722974582966214656   
1202  716439118184652801   
1228  713900603437621249   
1254  710658690886586372   
1274  709198395643068416   
1351  704054845121142784   
1433  697463031882764288   
1635  684222868335505415   
1662  682962037429899265   
1779  677716515794329600   
1843  675853064436391936   
2335  666287406224695296   

                                                                                                                                                                           text  \
433   The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd                                                                         
516   Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx   
902   Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE                                                                                              
1068  After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ                                
1120  Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv                                                   
1165  Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a                                                                                                            
1202  This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq                                                       
1228  Happy Saturday here's 9 puppers on a bench. 99/90 good work everybody https://t.co/mpvaVxKmc1                                                                               
1254  Here's a brigade of puppers. All look very prepared for whatever happens next. 80/80 https://t.co/0eb7R1Om12                                                                
1274  From left to right:\nCletus, Jerome, Alejandro, Burp, &amp; Titson\nNone know where camera is. 45/50 would hug all at once https://t.co/sedre1ivTK                          
1351  Here is a whole flock of puppers.  60/50 I'll take the lot https://t.co/9dpcw6MdWa                                                                                          
1433  Happy Wednesday here's a bucket of pups. 44/40 would pet all at once https://t.co/HppvrYuamZ                                                                                
1635  Someone help the girl is being mugged. Several are distracting her while two steal her shoes. Clever puppers 121/110 https://t.co/1zfnTJLt55                                
1662  This is Darrel. He just robbed a 7/11 and is in a high speed police chase. Was just spotted by the helicopter 10/10 https://t.co/7EsP8LmSp5                                 
1779  IT'S PUPPERGEDDON. Total of 144/120 ...I think https://t.co/ZanVtAtvIq                                                                                                      
1843  Here we have an entire platoon of puppers. Total score: 88/80 would pet all at once https://t.co/y93p6FLvVw                                                                 
2335  This is an Albanian 3 1/2 legged  Episcopalian. Loves well-polished hardwood flooring. Penis on the collar. 9/10 https://t.co/d9NcXFKwLv                                    

      rating_numerator  rating_denominator  
433   84                70                  
516   24                7                   
902   165               150                 
1068  9                 11                  
1120  204               170                 
1165  4           

tweet_id : 666287406224695296 score 1/2 instead of 9/10

tweet_id : 740373189193256964 score 9/11 instead of 14/10

tweet_id : 682962037429899265 score 7/11 instead of 10/10

tweet_id : 716439118184652801 score 50/50 instead of 11/10

tweet_id : 722974582966214656 score 4/20 instead of 13/10

In [30]:
twitter_archive_clean.loc[twitter_archive_clean.tweet_id == 666287406224695296, ['rating_numerator','rating_denominator']] = 9, 10
twitter_archive_clean.loc[twitter_archive_clean.tweet_id == 740373189193256964, ['rating_numerator','rating_denominator']] = 14, 10
twitter_archive_clean.loc[twitter_archive_clean.tweet_id == 682962037429899265, ['rating_numerator','rating_denominator']] = 10, 10
twitter_archive_clean.loc[twitter_archive_clean.tweet_id == 716439118184652801, ['rating_numerator','rating_denominator']] = 11, 10
twitter_archive_clean.loc[twitter_archive_clean.tweet_id == 722974582966214656, ['rating_numerator','rating_denominator']] = 13, 10

In [31]:
twitter_score_status2 = twitter_archive_clean[['tweet_id','text','rating_numerator','rating_denominator']]
pd.set_option('display.max_colwidth', -1)
twitter_score_status2.query("rating_denominator != 10")

tweet_id  \
433   820690176645140481   
516   810984652412424192   
902   758467244762497024   
1120  731156023742988288   
1228  713900603437621249   
1254  710658690886586372   
1274  709198395643068416   
1351  704054845121142784   
1433  697463031882764288   
1635  684222868335505415   
1779  677716515794329600   
1843  675853064436391936   

                                                                                                                                                                           text  \
433   The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd                                                                         
516   Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx   
902   Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE                                                                                              
1120  Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv                                                   
1228  Happy Saturday here's 9 puppers on a bench. 99/90 good work everybody https://t.co/mpvaVxKmc1                                                                               
1254  Here's a brigade of puppers. All look very prepared for whatever happens next. 80/80 https://t.co/0eb7R1Om12                                                                
1274  From left to right:\nCletus, Jerome, Alejandro, Burp, &amp; Titson\nNone know where camera is. 45/50 would hug all at once https://t.co/sedre1ivTK                          
1351  Here is a whole flock of puppers.  60/50 I'll take the lot https://t.co/9dpcw6MdWa                                                                                          
1433  Happy Wednesday here's a bucket of pups. 44/40 would pet all at once https://t.co/HppvrYuamZ                                                                                
1635  Someone help the girl is being mugged. Several are distracting her while two steal her shoes. Clever puppers 121/110 https://t.co/1zfnTJLt55                                
1779  IT'S PUPPERGEDDON. Total of 144/120 ...I think https://t.co/ZanVtAtvIq                                                                                                      
1843  Here we have an entire platoon of puppers. Total score: 88/80 would pet all at once https://t.co/y93p6FLvVw                                                                 

      rating_numerator  rating_denominator  
433   84                70                  
516   24                7                   
902   165               150                 
1120  204               170                 
1228  99                90                  
1254  80                80                  
1274  45                50                  
1351  60                50                  
1433  44                40                  
1635  121               110                 
1779  144               120                 
1843  88                80

<a id='clean4'></a>
#### IV.Delete denominator > 10

As seen right before, some denominator that fall under the recording error aren't equal to 10. those tweet doesn't have appropriate ratings, we are going to delete them.

In [32]:
twitter_archive_clean = twitter_archive_clean.query("rating_denominator == 10")
twitter_archive_clean['rating_denominator'].value_counts()

10    2085
Name: rating_denominator, dtype: int64

<a id='clean5'></a>
#### V. Absurd numerator ratings left

Let's check for some absurd ratings in the numerator

In [33]:
tweet_num_test = twitter_archive_clean.query("rating_numerator > 10")
tweet_num_test['rating_numerator'].value_counts()

12      486
11      414
13      288
14      39 
1776    1  
420     1  
75      1  
27      1  
26      1  
Name: rating_numerator, dtype: int64

Some numerator seems inconsistent let's check their ratings

In [34]:
tweet_num_test.query("rating_numerator > 25")

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
695   786709082849828864 NaN                    NaN                    
763   778027034220126208 NaN                    NaN                    
979   749981277374128128 NaN                    NaN                    
1712  680494726643068929 NaN                    NaN                    
2074  670842764863651840 NaN                    NaN                    

               timestamp  \
695  2016-10-13 23:23:56   
763  2016-09-20 00:24:34   
979  2016-07-04 15:00:45   
1712 2015-12-25 21:06:00   
2074 2015-11-29 05:52:33   

                                                                                   source  \
695   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>    
763   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>    
979   <a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>   
1712  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>    
2074  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>    

                                                                                                                                                                   text  \
695   This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wuqaPS                                    
763   This is Sophie. She's a Jubilant Bush Pupper. Super h*ckin rare. Appears at random just to smile at the locals. 11.27/10 would smile back https://t.co/QFaUiIHxHq   
979   This is Atticus. He's quite simply America af. 1776/10 https://t.co/GRXwMxLBkh                                                                                      
1712  Here we have uncovered an entire battalion of holiday puppers. Average of 11.26/10 https://t.co/eNm2S6p9BD                                                          
2074  After so many requests... here you go.\n\nGood dogg. 420/10 https://t.co/yfAAo1gdeY                                                                                 

      retweeted_status_id  retweeted_status_user_id  \
695  NaN                  NaN                         
763  NaN                  NaN                         
979  NaN                  NaN                         
1712 NaN                  NaN                         
2074 NaN                  NaN                         

     retweeted_status_timestamp  \
695   NaN                         
763   NaN                         
979   NaN                         
1712  NaN                         
2074  NaN                         

                                                        expanded_urls  \
695   https://twitter.com/dog_rates/status/786709082849828864/photo/1   
763   https://twitter.com/dog_rates/status/778027034220126208/photo/1   
979   https://twitter.com/dog_rates/status/749981277374128128/photo/1   
1712  https://twitter.com/dog_rates/status/680494726643068929/photo/1   
2074  https://twitter.com/dog_rates/status/670842764863651840/photo/1   

      rating_numerator  rating_denominator     name doggo floofer  pupper  \
695   75                10                  Logan    None  None    None     
763   27                10                  Sophie   None  None    pupper   
979   1776              10                  Atticus  None  None    None     
1712  26                10                  None     None  None    None     
2074  420               10                  None     None  None    None     

     puppo  
695   None  
763   None  
979   None  
1712  None  
2074  None

Due to decimal some rating were recorded incorrectly. let's fix this by correcting the numerator ratings to the nearest whole number.

In [35]:
twitter_archive_clean.loc[twitter_archive_clean.tweet_id == 786709082849828864, ['rating_numerator','rating_denominator']] = 10, 10
twitter_archive_clean.loc[twitter_archive_clean.tweet_id == 778027034220126208, ['rating_numerator','rating_denominator']] = 11, 10
twitter_archive_clean.loc[twitter_archive_clean.tweet_id == 680494726643068929, ['rating_numerator','rating_denominator']] = 11, 10

In [36]:
twitter_archive_clean['rating_numerator'].value_counts()

12      486
10      438
11      416
13      288
9       153
8       98 
7       51 
14      39 
5       34 
6       32 
3       19 
4       15 
2       9  
1       4  
420     1  
1776    1  
0       1  
Name: rating_numerator, dtype: int64

The ratings now seems more consistent, note that the scores of 1776 and 420 are not recording errors but the ratings given in the tweet.

<a id='clean6'></a>
#### VI. HTML in Source column

Using regex, replace the HTML tag

In [37]:
twitter_archive_clean.head(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193 NaN                    NaN                    
1  892177421306343426 NaN                    NaN                    
2  891815181378084864 NaN                    NaN                    

            timestamp  \
0 2017-08-01 16:23:56   
1 2017-08-01 00:17:27   
2 2017-07-31 00:18:03   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                 text  \
0  This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU                                                        
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2  This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB                    

   retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
0 NaN                  NaN                        NaN                         
1 NaN                  NaN                        NaN                         
2 NaN                  NaN                        NaN                         

                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/892420643555336193/photo/1   
1  https://twitter.com/dog_rates/status/892177421306343426/photo/1   
2  https://twitter.com/dog_rates/status/891815181378084864/photo/1   

   rating_numerator  rating_denominator     name doggo floofer pupper puppo  
0  13                10                  Phineas  None  None    None   None  
1  13                10                  Tilly    None  None    None   None  
2  12                10                  Archie   None  None    None   None

In [38]:
twitter_archive_clean['source'][0]

'<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>'

In [39]:
twitter_archive_clean['source'] = twitter_archive_clean['source'].str.replace('<[^<]+?>', '')

In [40]:
twitter_archive_clean['source'].value_counts()

Twitter for iPhone     1952
Vine - Make a Scene    91  
Twitter Web Client     31  
TweetDeck              11  
Name: source, dtype: int64

<a id='clean7'></a>
#### VII. Name column inconsistencies

it has been observed, some inconsistence in the name column.
I'll first detect them and replace to "None" and then delete all the None name value.

In [41]:
twitter_archive_clean['name'].value_counts()

None         594
a            54 
Charlie      11 
Lucy         11 
Cooper       10 
Oliver       10 
Penny        9  
Tucker       9  
Sadie        8  
the          8  
Lola         8  
Winston      8  
Toby         7  
Daisy        7  
Bo           6  
Bailey       6  
Jax          6  
Oscar        6  
an           6  
Koda         6  
Bella        6  
Stanley      6  
Buddy        5  
Louis        5  
Dave         5  
Rusty        5  
Scout        5  
Bentley      5  
Milo         5  
Leo          5  
            ..  
Clyde        1  
Anakin       1  
Tedders      1  
Willow       1  
Humphrey     1  
Goose        1  
Grizzwald    1  
Venti        1  
Vinscent     1  
Jareld       1  
Roscoe       1  
Cermet       1  
Tanner       1  
Sora         1  
General      1  
Philbert     1  
Kathmandu    1  
Koko         1  
BeBe         1  
Sunshine     1  
Lipton       1  
Pip          1  
Obi          1  
Taco         1  
Lolo         1  
Paull        1  
Ralphé       1  
Bauer        1

In [42]:
twit_fake_name = twitter_archive_clean['name'].str.contains('^[a-z]')
twitter_archive_clean[twit_fake_name].name.value_counts()

a               54
the             8 
an              6 
one             4 
very            4 
just            3 
quite           3 
getting         2 
actually        2 
not             2 
space           1 
mad             1 
old             1 
such            1 
my              1 
unacceptable    1 
all             1 
his             1 
officially      1 
infuriating     1 
incredibly      1 
by              1 
light           1 
life            1 
Name: name, dtype: int64

Except the None value who is capitalized, all the other small caps name are inconsistent

In [43]:
fk_name = ['a','the', 'an', 'very','one', 'quite', 'just', 'actually', 'getting', 'not', 'old','space', 'all', 'this', 'my', 'such', 'unacceptable', 'officially', 'by', 'light' ,'incredibly', 'his','mad', 'life', 'infuriating']

for names in twitter_archive_clean['name']:
    if names in fk_name:
        twitter_archive_clean['name'][twitter_archive_clean['name'] == names] = 'None'

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """


In [44]:
twitter_archive_clean['name'].value_counts()

None         696
Lucy         11 
Charlie      11 
Oliver       10 
Cooper       10 
Penny        9  
Tucker       9  
Lola         8  
Sadie        8  
Winston      8  
Daisy        7  
Toby         7  
Bailey       6  
Bella        6  
Koda         6  
Jax          6  
Stanley      6  
Oscar        6  
Bo           6  
Milo         5  
Bentley      5  
Rusty        5  
Louis        5  
Chester      5  
Leo          5  
Dave         5  
Scout        5  
Buddy        5  
Sophie       4  
Chip         4  
            ..  
Kota         1  
Harper       1  
Harnold      1  
Godzilla     1  
Clyde        1  
Mike         1  
Bauer        1  
Ralphé       1  
General      1  
Jay          1  
Clarq        1  
Grizzwald    1  
Vinscent     1  
Jareld       1  
Roscoe       1  
Cermet       1  
Tanner       1  
Sora         1  
Philbert     1  
Paull        1  
Kathmandu    1  
Koko         1  
BeBe         1  
Sunshine     1  
Lipton       1  
Pip          1  
Obi          1  
Taco         1

In [45]:
twitter_archive_clean = twitter_archive_clean.query("name !='None'")
twitter_archive_clean['name'].value_counts()

Charlie      11
Lucy         11
Cooper       10
Oliver       10
Penny        9 
Tucker       9 
Sadie        8 
Lola         8 
Winston      8 
Daisy        7 
Toby         7 
Bailey       6 
Stanley      6 
Jax          6 
Bella        6 
Koda         6 
Bo           6 
Oscar        6 
Buddy        5 
Leo          5 
Milo         5 
Scout        5 
Chester      5 
Louis        5 
Dave         5 
Bentley      5 
Rusty        5 
Larry        4 
Reggie       4 
Archie       4 
            .. 
Harper       1 
Harnold      1 
Godzilla     1 
Clyde        1 
Anakin       1 
Venti        1 
Mike         1 
Bauer        1 
Philbert     1 
Clarq        1 
Grizzwald    1 
Vinscent     1 
Jareld       1 
Roscoe       1 
Cermet       1 
Tanner       1 
Sora         1 
General      1 
Kathmandu    1 
Ralphé       1 
Koko         1 
BeBe         1 
Sunshine     1 
Lipton       1 
Pip          1 
Obi          1 
Taco         1 
Lolo         1 
Paull        1 
Willem       1 
Name: name, Length: 929,

All the inconsistent value in name column has been deleted

### `img_pred` table

In [46]:
image_prediction_clean = img_pred.copy()

In [47]:
image_prediction_clean.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0  1        Welsh_springer_spaniel  0.465074  True    collie               
1  1        redbone                 0.506826  True    miniature_pinscher   
2  1        German_shepherd         0.596461  True    malinois             
3  1        Rhodesian_ridgeback     0.408143  True    redbone              
4  1        miniature_pinscher      0.560311  True    Rottweiler           

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665  True    Shetland_sheepdog    0.061428  True    
1  0.074192  True    Rhodesian_ridgeback  0.072010  True    
2  0.138584  True    bloodhound           0.116197  True    
3  0.360687  True    miniature_pinscher   0.222752  True    
4  0.243682  True    Doberman             0.154629  True

<a id='clean8'></a>
#### VIII. standardization of dog breed names 

##### Phase 1 : Replace Undersore

In order to have standardized dog breed names we are going to replace the undersore per a space
Using the replace method in p1, p2 & p3 column

In [48]:
for breed in image_prediction_clean[['p1','p2','p3']]:
    image_prediction_clean[breed] = image_prediction_clean[breed].str.replace('_',' ')

In [49]:
image_prediction_clean.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0  1        Welsh springer spaniel  0.465074  True    collie               
1  1        redbone                 0.506826  True    miniature pinscher   
2  1        German shepherd         0.596461  True    malinois             
3  1        Rhodesian ridgeback     0.408143  True    redbone              
4  1        miniature pinscher      0.560311  True    Rottweiler           

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665  True    Shetland sheepdog    0.061428  True    
1  0.074192  True    Rhodesian ridgeback  0.072010  True    
2  0.138584  True    bloodhound           0.116197  True    
3  0.360687  True    miniature pinscher   0.222752  True    
4  0.243682  True    Doberman             0.154629  True

##### Phase 2 : capital letter for each words first letter

in the continuation of the objective of standardization of dog breed names we are going to replace in capital letter the first letter of each word present in the columns p1 p2 and p3.

Using the title() method

In [50]:
for breedcap in image_prediction_clean[['p1','p2','p3']]:
    image_prediction_clean[breedcap] = image_prediction_clean[breedcap].str.title()

In [51]:
image_prediction_clean.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0  1        Welsh Springer Spaniel  0.465074  True    Collie               
1  1        Redbone                 0.506826  True    Miniature Pinscher   
2  1        German Shepherd         0.596461  True    Malinois             
3  1        Rhodesian Ridgeback     0.408143  True    Redbone              
4  1        Miniature Pinscher      0.560311  True    Rottweiler           

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665  True    Shetland Sheepdog    0.061428  True    
1  0.074192  True    Rhodesian Ridgeback  0.072010  True    
2  0.138584  True    Bloodhound           0.116197  True    
3  0.360687  True    Miniature Pinscher   0.222752  True    
4  0.243682  True    Doberman             0.154629  True

<a id='tidy'></a>
### II. Tydinnes

<a id='tidy1'></a>
#### I. Stage dog column

Melt the *doggo*, *floofer*, *pupper*, *puppo* columns to a *stage* column.

In [52]:
twitter_archive_clean.head(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193 NaN                    NaN                    
1  892177421306343426 NaN                    NaN                    
2  891815181378084864 NaN                    NaN                    

            timestamp              source  \
0 2017-08-01 16:23:56  Twitter for iPhone   
1 2017-08-01 00:17:27  Twitter for iPhone   
2 2017-07-31 00:18:03  Twitter for iPhone   

                                                                                                                                                                 text  \
0  This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU                                                        
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2  This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB                    

   retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
0 NaN                  NaN                        NaN                         
1 NaN                  NaN                        NaN                         
2 NaN                  NaN                        NaN                         

                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/892420643555336193/photo/1   
1  https://twitter.com/dog_rates/status/892177421306343426/photo/1   
2  https://twitter.com/dog_rates/status/891815181378084864/photo/1   

   rating_numerator  rating_denominator     name doggo floofer pupper puppo  
0  13                10                  Phineas  None  None    None   None  
1  13                10                  Tilly    None  None    None   None  
2  12                10                  Archie   None  None    None   None

In [53]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1389 entries, 0 to 2325
Data columns (total 17 columns):
tweet_id                      1389 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     1389 non-null datetime64[ns]
source                        1389 non-null object
text                          1389 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 1389 non-null object
rating_numerator              1389 non-null int64
rating_denominator            1389 non-null int64
name                          1389 non-null object
doggo                         1389 non-null object
floofer                       1389 non-null object
pupper                        1389 non-null object
puppo                         1389 non-null object
dtypes: datetime64[ns](1), float

In [54]:
twitter_archive_clean2 = pd.melt(twitter_archive_clean, id_vars=['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp', 'source', 'text', 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp', 'expanded_urls', 'rating_numerator', 'rating_denominator', 'name'])
twitter_archive_clean2

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193 NaN                    NaN                    
1     892177421306343426 NaN                    NaN                    
2     891815181378084864 NaN                    NaN                    
3     891689557279858688 NaN                    NaN                    
4     891327558926688256 NaN                    NaN                    
5     890971913173991426 NaN                    NaN                    
6     890609185150312448 NaN                    NaN                    
7     890240255349198849 NaN                    NaN                    
8     890006608113172480 NaN                    NaN                    
9     889880896479866881 NaN                    NaN                    
10    889638837579907072 NaN                    NaN                    
11    889531135344209921 NaN                    NaN                    
12    889278841981685760 NaN                    NaN                    
13    888917238123831296 NaN                    NaN                    
14    888804989199671297 NaN                    NaN                    
15    888554962724278272 NaN                    NaN                    
16    888078434458587136 NaN                    NaN                    
17    887705289381826560 NaN                    NaN                    
18    887473957103951883 NaN                    NaN                    
19    886983233522544640 NaN                    NaN                    
20    886736880519319552 NaN                    NaN                    
21    886680336477933568 NaN                    NaN                    
22    886366144734445568 NaN                    NaN                    
23    886258384151887873 NaN                    NaN                    
24    885984800019947520 NaN                    NaN                    
25    885528943205470208 NaN                    NaN                    
26    884925521741709313 NaN                    NaN                    
27    884876753390489601 NaN                    NaN                    
28    884562892145688576 NaN                    NaN                    
29    884162670584377345 NaN                    NaN                    
...                  ...  ..                     ..                    
5526  667517642048163840 NaN                    NaN                    
5527  667502640335572993 NaN                    NaN                    
5528  667495797102141441 NaN                    NaN                    
5529  667455448082227200 NaN                    NaN                    
5530  667453023279554560 NaN                    NaN                    
5531  667405339315146752 NaN                    NaN                    
5532  667393430834667520 NaN                    NaN                    
5533  667211855547486208 NaN                    NaN                    
5534  667200525029539841 NaN                    NaN                    
5535  667182792070062081 NaN                    NaN                    
5536  667174963120574464 NaN                    NaN                    
5537  667171260800061440 NaN                    NaN                    
5538  667165590075940865 NaN                    NaN                    
5539  667160273090932737 NaN                    NaN                    
5540  667152164079423490 NaN                    NaN                    
5541  667119796878725120 NaN                    NaN                    
5542  667090893657276420 NaN                    NaN                    
5543  667073648344346624 NaN                    NaN                    
5544  667062181243039745 NaN                    NaN                    
5545  667012601033924608 NaN                    NaN                    
5546  666996132027977728 NaN                    NaN                    
5547  666817836334096384 NaN                    NaN                    
5548  666804364988780544 NaN                    NaN                    
5549  666776908487630848 NaN                    NaN     

In [55]:
stage_match = twitter_archive_clean2.query("variable == value")
stage_nomatch = twitter_archive_clean2.query("variable != value")

In [56]:
del stage_nomatch['variable']
del stage_match['variable']

In [57]:
stage_nomatch = stage_nomatch.drop_duplicates()
stage_match = stage_match.drop_duplicates()

In [58]:
stage_match.tweet_id.value_counts()

817777686764523521    2
801115127852503040    2
759793422261743616    2
785639753186217984    2
751950017322246144    1
681891461017812993    1
673697980713705472    1
685307451701334016    1
675006312288268288    1
686618349602762752    1
679503373272485890    1
828381636999917570    1
673919437611909120    1
819952236453363712    1
819004803107983360    1
811386762094317568    1
705975130514706432    1
750011400160841729    1
693095443459342336    1
826848821049180160    1
683142553609318400    1
675845657354215424    1
680473011644985345    1
686683045143953408    1
721001180231503872    1
867421006826221569    1
890240255349198849    1
703611486317502464    1
679777920601223168    1
760893934457552897    1
                     ..
878776093423087618    1
750868782890057730    1
719704490224398336    1
694329668942569472    1
681694085539872773    1
789903600034189313    1
871515927908634625    1
873213775632977920    1
781655249211752448    1
673612854080196609    1
8130969848233492

Most of the stage are indicated as None
4 dogs have 2 stages : 
                     -   817777686764523521 -> pupper       
                     -   801115127852503040 -> doggo
                     -   759793422261743616 -> doggo
                     -   785639753186217984 -> doggo


In [59]:
stage_match = stage_match.drop(270)
stage_match = stage_match.drop(3117)
stage_match = stage_match.drop(3311)
stage_match = stage_match.drop(3191)

In [60]:
twitter_archive_clean3 = pd.concat([stage_match, stage_nomatch])
twitter_archive_clean3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1573 entries, 7 to 2053
Data columns (total 14 columns):
tweet_id                      1573 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     1573 non-null datetime64[ns]
source                        1573 non-null object
text                          1573 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 1573 non-null object
rating_numerator              1573 non-null int64
rating_denominator            1573 non-null int64
name                          1573 non-null object
value                         1573 non-null object
dtypes: datetime64[ns](1), float64(4), int64(3), object(6)
memory usage: 184.3+ KB


In [61]:
twitter_archive_clean3.tweet_id.value_counts()

749417653287129088    2
801115127852503040    2
819006400881917954    2
711008018775851008    2
689623661272240129    2
845306882940190720    2
672898206762672129    2
683481228088049664    2
786363235746385920    2
692417313023332352    2
884162670584377345    2
672481316919734272    2
692905862751522816    2
674737130913071104    2
685169283572338688    2
680145970311643136    2
681694085539872773    2
789903600034189313    2
750383411068534784    2
689905486972461056    2
832032802820481025    2
760893934457552897    2
673707060090052608    2
742150209887731712    2
821407182352777218    2
696713835009417216    2
672594978741354496    2
800388270626521089    2
826958653328592898    2
869596645499047938    2
                     ..
667453023279554560    1
741099773336379392    1
745712589599014916    1
708853462201716736    1
780858289093574656    1
735991953473572864    1
675111688094527488    1
669000397445533696    1
797236660651966464    1
777885040357281792    1
8258765121591869

In [62]:
twitter_archive_clean3 = twitter_archive_clean3.drop_duplicates(subset='tweet_id', keep="first")

In [63]:
twitter_archive_clean3.rename(columns={'value': 'stage'}, inplace=True)

In [64]:
twitter_archive_clean3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1389 entries, 7 to 1388
Data columns (total 14 columns):
tweet_id                      1389 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     1389 non-null datetime64[ns]
source                        1389 non-null object
text                          1389 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 1389 non-null object
rating_numerator              1389 non-null int64
rating_denominator            1389 non-null int64
name                          1389 non-null object
stage                         1389 non-null object
dtypes: datetime64[ns](1), float64(4), int64(3), object(6)
memory usage: 162.8+ KB


In [65]:
twitter_archive_clean = twitter_archive_clean3

<a id='tidy2'></a>
#### II. Delete the retweet and reply columns from archive

In [66]:
del twitter_archive_clean['in_reply_to_status_id']
del twitter_archive_clean['in_reply_to_user_id']
del twitter_archive_clean['retweeted_status_user_id']
del twitter_archive_clean['retweeted_status_id']
del twitter_archive_clean['retweeted_status_timestamp']

In [67]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1389 entries, 7 to 1388
Data columns (total 9 columns):
tweet_id              1389 non-null int64
timestamp             1389 non-null datetime64[ns]
source                1389 non-null object
text                  1389 non-null object
expanded_urls         1389 non-null object
rating_numerator      1389 non-null int64
rating_denominator    1389 non-null int64
name                  1389 non-null object
stage                 1389 non-null object
dtypes: datetime64[ns](1), int64(3), object(5)
memory usage: 108.5+ KB


<a id='tidy3'></a>
#### III. Tydinnes issues of image_prediction_clean

Here we have 9 Columns to represent the 3 predictions,
I would like to replace them in 4 columns to select only the best prediction and know which one it was.
the four columns would be as follows: 
    - # predictions
    - dog breed name
    - confidence of prediction
    - boolean of whether it's a dog or not.

In [68]:
image_prediction_clean.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0  1        Welsh Springer Spaniel  0.465074  True    Collie               
1  1        Redbone                 0.506826  True    Miniature Pinscher   
2  1        German Shepherd         0.596461  True    Malinois             
3  1        Rhodesian Ridgeback     0.408143  True    Redbone              
4  1        Miniature Pinscher      0.560311  True    Rottweiler           

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665  True    Shetland Sheepdog    0.061428  True    
1  0.074192  True    Rhodesian Ridgeback  0.072010  True    
2  0.138584  True    Bloodhound           0.116197  True    
3  0.360687  True    Miniature Pinscher   0.222752  True    
4  0.243682  True    Doberman             0.154629  True

In [69]:
image_prediction_clean2 =  image_prediction_clean
image_prediction_clean2.head(3)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0  1        Welsh Springer Spaniel  0.465074  True    Collie               
1  1        Redbone                 0.506826  True    Miniature Pinscher   
2  1        German Shepherd         0.596461  True    Malinois             

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665  True    Shetland Sheepdog    0.061428  True    
1  0.074192  True    Rhodesian Ridgeback  0.072010  True    
2  0.138584  True    Bloodhound           0.116197  True

In [70]:
max_conf = image_prediction_clean2[["p1_conf", "p2_conf","p3_conf"]].max(axis=1)
image_prediction_clean2['confidence'] = max_conf

In [71]:
image_prediction_clean2

tweet_id  \
0     666020888022790149   
1     666029285002620928   
2     666033412701032449   
3     666044226329800704   
4     666049248165822465   
5     666050758794694657   
6     666051853826850816   
7     666055525042405380   
8     666057090499244032   
9     666058600524156928   
10    666063827256086533   
11    666071193221509120   
12    666073100786774016   
13    666082916733198337   
14    666094000022159362   
15    666099513787052032   
16    666102155909144576   
17    666104133288665088   
18    666268910803644416   
19    666273097616637952   
20    666287406224695296   
21    666293911632134144   
22    666337882303524864   
23    666345417576210432   
24    666353288456101888   
25    666362758909284353   
26    666373753744588802   
27    666396247373291520   
28    666407126856765440   
29    666411507551481857   
...                  ...   
2045  886366144734445568   
2046  886680336477933568   
2047  886736880519319552   
2048  886983233522544640   
2049  887101392804085760   
2050  887343217045368832   
2051  887473957103951883   
2052  887517139158093824   
2053  887705289381826560   
2054  888078434458587136   
2055  888202515573088257   
2056  888554962724278272   
2057  888804989199671297   
2058  888917238123831296   
2059  889278841981685760   
2060  889531135344209921   
2061  889638837579907072   
2062  889665388333682689   
2063  889880896479866881   
2064  890006608113172480   
2065  890240255349198849   
2066  890609185150312448   
2067  890729181411237888   
2068  890971913173991426   
2069  891087950875897856   
2070  891327558926688256   
2071  891689557279858688   
2072  891815181378084864   
2073  892177421306343426   
2074  892420643555336193   

                                                                                      jpg_url  \
0     https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg                                           
1     https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg                                           
2     https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg                                           
3     https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg                                           
4     https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg                                           
5     https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg                                           
6     https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg                                           
7     https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg                                           
8     https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg                                           
9     https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg                                           
10    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg                                           
11    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg                                           
12    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg                                           
13    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg                                           
14    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg                                           
15    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg                                           
16    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg                                           
17    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg                                           
18    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg                                           
19    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg                                           
20    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg                                           
21    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg                                           
22    https://pbs.twimg.com/media/CT9OwFIWEAMuRj

let's see if the first prediction is always the better

In [72]:
image_prediction_clean2['check'] = (image_prediction_clean2['p1_conf'] == image_prediction_clean2['confidence']).astype('object')

In [73]:
image_prediction_clean2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 14 columns):
tweet_id      2075 non-null int64
jpg_url       2075 non-null object
img_num       2075 non-null int64
p1            2075 non-null object
p1_conf       2075 non-null float64
p1_dog        2075 non-null bool
p2            2075 non-null object
p2_conf       2075 non-null float64
p2_dog        2075 non-null bool
p3            2075 non-null object
p3_conf       2075 non-null float64
p3_dog        2075 non-null bool
confidence    2075 non-null float64
check         2075 non-null object
dtypes: bool(3), float64(4), int64(2), object(5)
memory usage: 184.5+ KB


In [74]:
image_prediction_clean2['check'].value_counts()

True    2075
Name: check, dtype: int64

The first prediction is always the better

In [75]:
image_prediction_clean2['breed'] = image_prediction_clean2.iloc[:,[3]]
image_prediction_clean2['isdog?'] = image_prediction_clean2.iloc[:,[5]]
del image_prediction_clean2['check']

In [76]:
image_prediction_clean2 = image_prediction_clean2.drop(['p1', 'p2','p3','p1_conf', 'p2_conf','p3_conf','p1_dog', 'p2_dog','p3_dog'], axis = 1)

In [77]:
image_prediction_clean2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 6 columns):
tweet_id      2075 non-null int64
jpg_url       2075 non-null object
img_num       2075 non-null int64
confidence    2075 non-null float64
breed         2075 non-null object
isdog?        2075 non-null bool
dtypes: bool(1), float64(1), int64(2), object(2)
memory usage: 83.2+ KB


In [78]:
image_prediction_clean = image_prediction_clean2

<a id='merge'></a>
### Merging column

In [79]:
image_prediction_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 6 columns):
tweet_id      2075 non-null int64
jpg_url       2075 non-null object
img_num       2075 non-null int64
confidence    2075 non-null float64
breed         2075 non-null object
isdog?        2075 non-null bool
dtypes: bool(1), float64(1), int64(2), object(2)
memory usage: 83.2+ KB


In [80]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1389 entries, 7 to 1388
Data columns (total 9 columns):
tweet_id              1389 non-null int64
timestamp             1389 non-null datetime64[ns]
source                1389 non-null object
text                  1389 non-null object
expanded_urls         1389 non-null object
rating_numerator      1389 non-null int64
rating_denominator    1389 non-null int64
name                  1389 non-null object
stage                 1389 non-null object
dtypes: datetime64[ns](1), int64(3), object(5)
memory usage: 108.5+ KB


In [81]:
tweet_metrics.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 3 columns):
tweet_id    2331 non-null object
fav         2331 non-null int64
rt          2331 non-null int64
dtypes: int64(2), object(1)
memory usage: 54.7+ KB


In [82]:
tweet_metrics['tweet_id'] = tweet_metrics['tweet_id'].astype(str).astype(int)

In [83]:
tweet_metrics.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 3 columns):
tweet_id    2331 non-null int64
fav         2331 non-null int64
rt          2331 non-null int64
dtypes: int64(3)
memory usage: 54.7 KB


In [84]:
twitter_archive_master = pd.merge(twitter_archive_clean, image_prediction_clean, on = 'tweet_id', how = 'left')

In [87]:
twitter_archive_master = pd.merge(twitter_archive_master, tweet_metrics, on = 'tweet_id', how = 'left')

In [88]:
twitter_archive_master

tweet_id           timestamp               source  \
0     890240255349198849 2017-07-26 15:59:51  Twitter for iPhone    
1     884162670584377345 2017-07-09 21:29:42  Twitter for iPhone    
2     871515927908634625 2017-06-04 23:56:03  Twitter for iPhone    
3     869596645499047938 2017-05-30 16:49:31  Twitter for iPhone    
4     846514051647705089 2017-03-28 00:07:32  Twitter for iPhone    
5     845397057150107648 2017-03-24 22:08:59  Twitter for iPhone    
6     836753516572119041 2017-03-01 01:42:39  Twitter for iPhone    
7     832998151111966721 2017-02-18 17:00:10  Twitter for iPhone    
8     832273440279240704 2017-02-16 17:00:25  Twitter for iPhone    
9     832032802820481025 2017-02-16 01:04:13  Twitter for iPhone    
10    831939777352105988 2017-02-15 18:54:34  Twitter for iPhone    
11    831322785565769729 2017-02-14 02:02:51  Twitter for iPhone    
12    829374341691346946 2017-02-08 17:00:26  Twitter for iPhone    
13    828381636999917570 2017-02-05 23:15:47  Twitter for iPhone    
14    826958653328592898 2017-02-02 01:01:21  Twitter for iPhone    
15    826848821049180160 2017-02-01 17:44:55  Twitter for iPhone    
16    826476773533745153 2017-01-31 17:06:32  Twitter for iPhone    
17    821765923262631936 2017-01-18 17:07:18  Twitter for iPhone    
18    821407182352777218 2017-01-17 17:21:47  Twitter for iPhone    
19    819006400881917954 2017-01-11 02:21:57  Twitter for iPhone    
20    819004803107983360 2017-01-11 02:15:36  Twitter for iPhone    
21    813910438903693312 2016-12-28 00:52:25  Twitter for iPhone    
22    813096984823349248 2016-12-25 19:00:02  Twitter for iPhone    
23    801115127852503040 2016-11-22 17:28:25  Twitter for iPhone    
24    799757965289017345 2016-11-18 23:35:32  Twitter for iPhone    
25    786363235746385920 2016-10-13 00:29:39  Twitter for iPhone    
26    785639753186217984 2016-10-11 00:34:48  Twitter for iPhone    
27    782747134529531904 2016-10-03 01:00:34  Twitter for iPhone    
28    782305867769217024 2016-10-01 19:47:08  Twitter for iPhone    
29    781655249211752448 2016-09-30 00:41:48  Vine - Make a Scene   
...                  ...                 ...                  ...   
1359  667517642048163840 2015-11-20 01:39:42  Twitter Web Client    
1360  667502640335572993 2015-11-20 00:40:05  Twitter Web Client    
1361  667495797102141441 2015-11-20 00:12:54  Twitter Web Client    
1362  667455448082227200 2015-11-19 21:32:34  Twitter Web Client    
1363  667453023279554560 2015-11-19 21:22:56  Twitter Web Client    
1364  667405339315146752 2015-11-19 18:13:27  Twitter for iPhone    
1365  667393430834667520 2015-11-19 17:26:08  Twitter for iPhone    
1366  667211855547486208 2015-11-19 05:24:37  Twitter for iPhone    
1367  667200525029539841 2015-11-19 04:39:35  Twitter for iPhone    
1368  667182792070062081 2015-11-19 03:29:07  Twitter for iPhone    
1369  667174963120574464 2015-11-19 02:58:01  Twitter for iPhone    
1370  667171260800061440 2015-11-19 02:43:18  Twitter for iPhone    
1371  667165590075940865 2015-11-19 02:20:46  Twitter for iPhone    
1372  667160273090932737 2015-11-19 01:59:39  Twitter for iPhone    
1373  667152164079423490 2015-11-19 01:27:25  Twitter for iPhone    
1374  667119796878725120 2015-11-18 23:18:48  Twitter for iPhone    
1375  667090893657276420 2015-11-18 21:23:57  Twitter for iPhone    
1376  667073648344346624 2015-11-18 20:15:26  Twitter for iPhone    
1377  667062181243039745 2015-11-18 19:29:52  Twitter for iPhone    
1378  667012601033924608 2015-11-18 16:12:51  Twitter for iPhone    
1379  666996132027977728 2015-11-18 15:07:24  Twitter for iPhone    
1380  666817836334096384 2015-11-18 03:18:55  Twitter for iPhone    
1381  666804364988780544 2015-11-18 02:25:23  Twitter for iPhone    
1382  666776908487630848 2015-11-18 00:36:17  Twitter for iPhone    
1383  666739327293083650 2015-11-17 22:06:57  Twitter for iPhone    
1384  666691418707132416 2015-11-17 18:56:35  Twitter for iPhone    
1385  66664482

In [89]:
twitter_archive_master = twitter_archive_master.drop(['jpg_url', 'img_num', 'fav_y', 'rt_y'], axis = 1)
twitter_archive_master.rename(columns={'fav_x':'favorite', 'rt_x':'retweet'}, inplace=True)

Let's check if the twitter_archive_master table is clean

In [91]:
twitter_archive_master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1389 entries, 0 to 1388
Data columns (total 14 columns):
tweet_id              1389 non-null int64
timestamp             1389 non-null datetime64[ns]
source                1389 non-null object
text                  1389 non-null object
expanded_urls         1389 non-null object
rating_numerator      1389 non-null int64
rating_denominator    1389 non-null int64
name                  1389 non-null object
stage                 1389 non-null object
confidence            1348 non-null float64
breed                 1348 non-null object
isdog?                1348 non-null object
favorite              1382 non-null float64
retweet               1382 non-null float64
dtypes: datetime64[ns](1), float64(3), int64(3), object(7)
memory usage: 162.8+ KB


As expected the column merged in the prior step does not exactly retrieve all the tweet_id.
We're gonna to delete the missing value of the last 5 columns

In [94]:
twitter_archive_master = twitter_archive_master.dropna(axis=0, subset=['confidence'])
twitter_archive_master = twitter_archive_master.dropna(axis=0, subset=['favorite'])

In [95]:
twitter_archive_master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1341 entries, 0 to 1388
Data columns (total 14 columns):
tweet_id              1341 non-null int64
timestamp             1341 non-null datetime64[ns]
source                1341 non-null object
text                  1341 non-null object
expanded_urls         1341 non-null object
rating_numerator      1341 non-null int64
rating_denominator    1341 non-null int64
name                  1341 non-null object
stage                 1341 non-null object
confidence            1341 non-null float64
breed                 1341 non-null object
isdog?                1341 non-null object
favorite              1341 non-null float64
retweet               1341 non-null float64
dtypes: datetime64[ns](1), float64(3), int64(3), object(7)
memory usage: 157.1+ KB


The twitter_archive_master seems now to be cleaner, it's time to store it into a csv

<a id='store'></a>
### Store twitter_archive_master

In [96]:
twitter_archive_master.to_csv('twitter_archive_master.csv', index = False)

# To see the Analysis project please go to Analysis_Act.ipynb